src/daft-csv/src/read.rs

```diff
@@ -199,6 +220,387 @@ fn tables_concat(mut tables: Vec<Table>) -> DaftResult<Table> {
+
+    let convert_options = match (convert_options, &predicate) {
+        (None, _) => None,
+        (co, None) => co,
+        (Some(mut co), Some(predicate)) => {
+            if let Some(ref mut include_columns) = co.include_columns {
+                let required_columns_for_predicate = get_required_columns(predicate);
+                for rc in required_columns_for_predicate {
+                    if include_columns.iter().all(|c| c.as_str() != rc.as_str()) {
+                        include_columns.push(rc)
+                    }
+                }
+            }
+            // If we have a limit and a predicate, remove limit for stream.
+            co.limit = None;
+            Some(co)
+        }
+    }
+    .unwrap_or_default();
+    // End of `should dedup`.
+
+    // TODO(desmond): We should do better schema inference here.
+    let schema = convert_options.clone().schema.unwrap().to_arrow()?;
+    let n_threads: usize = std::thread::available_parallelism()
+        .unwrap_or(NonZeroUsize::new(2).unwrap())
+        .into();
+    let chunk_size = read_options
+        .as_ref()
+        .and_then(|opt| opt.chunk_size.or_else(|| opt.buffer_size.map(|bs| bs / 8)))
+        .unwrap_or(DEFAULT_CHUNK_SIZE);
+    let projection_indices = fields_to_projection_indices(
+        &schema.clone().fields,
+        &convert_options.clone().include_columns,
+    );
+    let fields = schema.clone().fields;
+    let fields_subset = projection_indices
+        .iter()
+        .map(|i| fields.get(*i).unwrap().into())
+        .collect::<Vec<daft_core::datatypes::Field>>();
+    let read_schema = Arc::new(daft_core::schema::Schema::new(fields_subset)?);
+    let read_daft_fields = Arc::new(
+        read_schema
+            .fields
+            .values()
+            .map(|f| Arc::new(f.clone()))
+            .collect::<Vec<_>>(),
+    );
+    // TODO(desmond): Need better upfront estimators. Sample or keep running count of stats.
+    let estimated_mean_row_size = 100f64;
+    let estimated_std_row_size = 20f64;
+    let record_buffer_size = (estimated_mean_row_size + estimated_std_row_size).ceil() as usize;
+    let chunk_size_rows = (chunk_size as f64 / record_buffer_size as f64).ceil() as usize;
+    // TODO(desmond): We don't want to create a per-read buffer pool, we want one pool shared with
+    // the whole process.
+    let buffer_pool = CsvBufferPool::new(
+        record_buffer_size,
+        schema.fields.len(),
+        chunk_size_rows,
+        n_threads * 2,
+    );
+    let chunk_offsets: Vec<usize> = (0..=bytes.len()).step_by(chunk_size).collect();
+    // TODO(desmond): Memory usage is still growing during execution of a .count(*).collect(), so
+    // the following approach still isn't quite right.
+    // TODO(desmond): Also, is this usage of max_chunks_in_flight correct?
+    let (sender, receiver) =
+        crossbeam_channel::bounded(max_chunks_in_flight.unwrap_or(n_threads * 2));
```

Comment:
I think you might deadlock here, e.g. if your channel is full and all threads are trying to send chunks, they will all block. For parquet, we capped it to `expected_num_chunks`, i.e. the total amount of chunks that will ever be read. See: https://github.com/Eventual-Inc/Daft/pull/2620/files#diff-f65cce5db7fd92cf9fe1cc38ebd495165b7ac59444a91f0c93c4c4c027089429R392-R400

Reply:
Oh and also (maybe you have this as a todo but just in case), the ordering should also be preserved, (for the streaming executor, there's a `maintain_order` flag that indicates if order needs to be preserved, but the non-streaming one requires order). I don't think `into_par_iter.for_each` will preserve it but correct me if im wrong